### PR TITLE
[DEV-3483] Bump Pydantic to V2

### PR DESCRIPTION
**Pull request by gee-senbong**
## Description

This PR updates the following packages along with pydantic upgrade
* pydantic
* fastapi
* feast
* typeguard

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: Taskfile.yaml [hunk 1]
```diff
@@ -151,12 +151,13 @@ tasks:
...
       PYLINT_DISABLE_FOR_TESTS: 'redefined-outer-name,invalid-name,protected-access,too-few-public-methods,unspecified-encoding,duplicate-code'
+      PYLINT_PLUGINS: 'pylint_pydantic'
     sources:
       - featurebyte/**/*
       - tests/**/*
     cmds:
-      - poetry run pylint --disable={{ .PYLINT_DISABLE_FOR_SRCS }} --rcfile pyproject.toml featurebyte
-      - poetry run pylint --disable={{ .PYLINT_DISABLE_FOR_TESTS }} --rcfile pyproject.toml tests
+      - poetry run pylint --load-plugins {{ .PYLINT_PLUGINS }} --disable={{ .PYLINT_DISABLE_FOR_SRCS }} --rcfile pyproject.toml featurebyte
+      - poetry run pylint --load-plugins {{ .PYLINT_PLUGINS }} --disable={{ .PYLINT_DISABLE_FOR_TESTS }} --rcfile pyproject.toml tests
```
Comment (gee-senbong) [inline Taskfile.yaml:160]: Introduce `pylint_pydantic` plugin to handle pylint `no-member` errors when refering attribute of pydantic model.

File: featurebyte/_overrides/typechecked_override.py [hunk 2]
```diff
@@ -3,47 +3,70 @@
...
-    func: Any = None,
+    target: T_CallableOrType | None = None,
     *,
-    always: bool = False,
-    _localns: Optional[Dict[str, Any]] = None,
+    forward_ref_policy: ForwardRefPolicy | Unset = unset,
+    typecheck_fail_callback: TypeCheckFailCallback | Unset = unset,
+    collection_check_strategy: CollectionCheckStrategy | Unset = unset,
+    debug_instrumentation: bool | Unset = unset,
 ) -> Any:
```
Comment (gee-senbong) [inline featurebyte/_overrides/typechecked_override.py:27]: Update signature here as newer typeguard expects different function signature.

File: featurebyte/core/generic.py [hunk 3]
```diff
@@ -268,48 +269,7 @@ def dict(self, *args: Any, **kwargs: Any) -> dict[str, Any]:
...
-        exclude: AbstractSetIntStr | MappingIntStrAny | None = None,
-        by_alias: bool = False,
-        skip_defaults: bool | None = None,
-        exclude_unset: bool = False,
-        exclude_defaults: bool = False,
-        exclude_none: bool = False,
-        encoder: Callable[[Any], Any] | None = None,
-        models_as_dict: bool = True,
-        **dumps_kwargs: Any,
-    ) -> str:
```
Comment (gee-senbong) [inline featurebyte/core/generic.py:286]: Remove this as it is no longer an issue in pydantic V2.

File: featurebyte/feast/utils/registry_construction.py [hunk 4]
```diff
@@ -925,7 +926,6 @@ def create_feast_registry_proto_for_feature_materialization(
             feature_services=[],
             on_demand_feature_views=[],
             stream_feature_views=[],
-            request_feature_views=[],
         )
```
Comment (gee-senbong) [inline featurebyte/feast/utils/registry_construction.py:929]: Remove this attribute as it is no longer valid in the newer `RepoContents` (from `feast` package).

File: featurebyte/schema/worker/task/feature_list_create.py [hunk 5]
```diff
@@ -21,10 +22,35 @@ class FeatureParameters(FeatureByteBaseModel):
...
+
+
 class FeaturesParameters(FeatureByteBaseModel):
     """Feature list feature parameters"""
 
-    features: Union[List[FeatureParameters], List[PydanticObjectId]]
+    features: Union[
+        Annotated[List[FeatureParameters], Tag("feature_params")],
+        Annotated[List[PydanticObjectId], Tag("feature_ids")],
+    ] = Field(discriminator=Discriminator(feature_params_discriminator), min_length=1)
```
Comment (gee-senbong) [inline featurebyte/schema/worker/task/feature_list_create.py:53]: `features: Union[List[FeatureParameters], List[PydanticObjectId]]` is no longer work in pydantic V2. In pydantic V2, a discrimination function is required to deserialize.

File: featurebyte/core/generic.py [hunk 6]
```diff
@@ -148,13 +147,14 @@ def node_types_lineage(self) -> list[NodeType]:
         -------
         list[NodeType]
         """
+        # pylint: disable=no-member
```
Comment (YS-L) [inline featurebyte/core/generic.py:150]: Is this pylint disable still needed with the pylint plugin?

File: featurebyte/common/documentation/resource_extractor.py [hunk 7]
```diff
@@ -7,14 +7,16 @@
 from typing import Any, Dict, List, Optional, get_type_hints
 
 import inspect
+import pdb
```
Comment (YS-L) [inline featurebyte/common/documentation/resource_extractor.py:10]: Is it intended to have pdb imported?

File: featurebyte/common/documentation/gen_ref_pages_docs_builder.py
Comment (YS-L) [inline featurebyte/common/documentation/gen_ref_pages_docs_builder.py:1]: Are there any changes the generated documentation? I saw the tag was added but can't find the docs html comparison.

File: featurebyte/feast/model/online_store.py [hunk 8]
```diff
@@ -89,4 +102,12 @@ def get_feast_online_store_details(
...
-    return parse_obj_as(FeastOnlineStoreDetails, online_store_details.dict())  # type: ignore
+    params = online_store_details.dict(by_alias=True)
+    online_store_detail_class = ONLINE_STORE_DETAILS_CLASS_MAP.get(params.get("type"))
+    if online_store_detail_class is None:
+        # use pydantic builtin version to throw validation error (slow due to pydantic V2 performance issue)
+        return TypeAdapter(FeastOnlineStoreDetails).validate_python(params)
+
+    # use internal method to avoid current pydantic V2 performance issue due to _core_utils.py:walk
+    # https://github.com/pydantic/pydantic/issues/6768
+    return cast(FeastOnlineStoreDetails, online_store_detail_class(**params))
```
Comment (YS-L) [inline featurebyte/feast/model/online_store.py:113]: This logic of using class map to construct an object seems to be repeated in a few places. Would it make sense to extract a helper function for reuse?

File: featurebyte/models/base.py [hunk 9]
```diff
@@ -142,24 +112,12 @@ def json_dict(self, **kwargs: Any) -> dict[str, Any]:
...
-        Configurations for FeatureByteBaseModel
-        """
-
-        # With `validate_assignment` flag enabled, pydantic model runs validation check during attribute assignment.
-        # This also enables the feature to make attribute immutable by using Field(allow_mutation=False).
-        validate_assignment = True
-
-        # With `use_enum_values` flag enabled, pydantic model converts the enum attribute to the enum's value when
-        # storing the value in the model (`<model>.<enum_attribute>` should return enum's value rather than enum type).
-        use_enum_values = True
```
Comment (YS-L) [inline featurebyte/models/base.py:156]: Are these comments still valid? If so might want to consider keeping them.

File: featurebyte/models/feature.py [hunk 10]
```diff
@@ -71,51 +72,39 @@ class BaseFeatureModel(QueryGraphMixin, FeatureByteCatalogBaseDocumentModel):
     It contains all the attributes that are shared between FeatureModel & TargetModel.
     """
 
-    dtype: DBVarType = Field(allow_mutation=False, default=DBVarType.UNKNOWN)
+    dtype: DBVarType = Field(default=DBVarType.UNKNOWN.value)
```
Comment (YS-L) [inline featurebyte/models/feature.py:75]: Should this have `frozen=True`? Same question for some of the fields below that have `allow_mutation=False` before but without `frozen=True` after.

File: featurebyte/models/mixin.py [hunk 11]
```diff
@@ -20,9 +22,14 @@ class QueryGraphMixin(FeatureByteBaseModel):
...
     # special handling for those attributes that are expensive to deserialize
     # internal_* is used to store the raw data from persistence, _* is used as a cache
-    internal_graph: Any = Field(default=None, allow_mutation=False, alias="graph")
+    internal_graph: Any = Field(default=None, frozen=True, alias="graph")
     _graph: Optional[QueryGraph] = PrivateAttr(default=None)
 
+    @field_serializer("internal_graph", when_used="json")
+    def _serialize_graph(self, graph: Any) -> Any:
+        _ = graph
+        return json.loads(self.graph.model_dump_json(by_alias=True))
```
Comment (YS-L) [inline featurebyte/models/mixin.py:31]: To help my understanding, could you clarify what does a field serializer do and when should a model use one in our context?

File: featurebyte/core/generic.py (hunk 6 above)
Comment (gee-senbong) [reply featurebyte/core/generic.py:150]: Good point, this can be removed.

File: featurebyte/common/documentation/resource_extractor.py (hunk 7 above)
Comment (gee-senbong) [reply featurebyte/common/documentation/resource_extractor.py:10]: Good catch, will remove this import.

File: featurebyte/query_graph/model/graph.py [hunk 12]
```diff
@@ -54,11 +54,16 @@ class QueryGraphModel(FeatureByteBaseModel):
...
-        return self.json(by_alias=True, indent=4)
+        return self.model_dump_json(by_alias=True, indent=4)
 
     def __str__(self) -> str:
         return repr(self)
 
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, QueryGraphModel):
+            return bool(self.dict() == other.dict())
+        return False
```
Comment (YS-L) [inline featurebyte/query_graph/model/graph.py:65]: Curious when will we do equality comparison between two query graph models?

File: featurebyte/common/documentation/gen_ref_pages_docs_builder.py
Comment (gee-senbong) [reply featurebyte/common/documentation/gen_ref_pages_docs_builder.py:1]: Yes, there are some changes in the generated documentation. Will remove the tag to regenerate the documentation diff.

File: featurebyte/models/mixin.py (hunk 11 above)
Comment (gee-senbong) [reply featurebyte/models/mixin.py:31]: `json` output should not contain type like `ObjectId` and `datetime`. Pydantic V2 deprecates `json_encoders` due to performance issue. We need to declare an explicit `field_serializer` method to serialize a particular field.

File: featurebyte/routes/event_table/api.py [hunk 13]
```diff
@@ -145,7 +146,7 @@ async def get_event_table_info(
         """
         controller = self.get_controller_for_request(request)
         info = await controller.get_info(
-            document_id=event_table_id,
+            document_id=ObjectId(event_table_id),
```
Comment (YS-L) [inline featurebyte/routes/event_table/api.py:149]: Will we get type check or lint error if we forgot to apply this conversion here?

Comment (gee-senbong) [reply featurebyte/routes/event_table/api.py:149]: We will get type check error here though `event_table_id` is already an `ObjectId`. I think the reason is that `PyObjectId` is inherited `str` type.

**Review by YS-L — APPROVED** [review]

File: featurebyte/query_graph/model/graph.py (hunk 12 above)
Comment (gee-senbong) [reply featurebyte/query_graph/model/graph.py:65]: We mainly use the comparison in our tests.

**Comment by gee-senbong** [comment]
Summary of Change of documentation:
* Change from `PydanticObjectId` to `ObjectId` or `Annotated[ObjectId, _ObjectIdPydanticAnnotation]`
* Change from `NameStr` to `Annotated[str, StringConstraints(strip_whitespace=None, to_upper=None, to_lower=None, strict=None, min_length=0, max_length=255, pattern=None)]`
* Change from `StrictStr` to `str` or `Annotated[str, Strict(strict=True)]`
* Change from `Annotated[Union[RedisOnlineStoreDetails, MySQLOnlineStoreDetails]]` to `Union[RedisOnlineStoreDetails, MySQLOnlineStoreDetails]`
* Change from `Union[StrictInt, StrictFloat, StrictStr, bool]` to `Union[bool, int, float, str]`
* Change from `StorageCredentialType` to `Literal["GCS"]` or `Literal["S3"]` ...

**Pull request merged**